File: src/main/java/com/yupi/template/model/vo/BudgetStatusVO.java
package com.yupi.template.model.vo;

import lombok.Data;

import java.io.Serializable;
import java.math.BigDecimal;

/**
 * 预算状态
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@Data
public class BudgetStatusVO implements Serializable {

    /**
     * 是否可以继续使用
     */
    private Boolean canProceed;

    /**
     * 预算状态：normal-正常，warning-预警，exceeded-超出
     */
    private String status;

    /**
     * 提示消息
     */
    private String message;

    /**
     * 今日已消耗
     */
    private BigDecimal todayCost;

    /**
     * 本月已消耗
     */
    private BigDecimal monthCost;

    /**
     * 日预算限额
     */
    private BigDecimal dailyBudget;

    /**
     * 月预算限额
     */
    private BigDecimal monthlyBudget;

    /**
     * 日预算使用百分比
     */
    private BigDecimal dailyUsagePercent;

    /**
     * 月预算使用百分比
     */
    private BigDecimal monthlyUsagePercent;

    private static final long serialVersionUID = 1L;

    /**
     * 创建正常状态
     */
    public static BudgetStatusVO normal(BigDecimal todayCost, BigDecimal monthCost,
                                        BigDecimal dailyBudget, BigDecimal monthlyBudget) {
        BudgetStatusVO vo = new BudgetStatusVO();
        vo.setCanProceed(true);
        vo.setStatus("normal");
        vo.setMessage("预算充足");
        vo.setTodayCost(todayCost);
        vo.setMonthCost(monthCost);
        vo.setDailyBudget(dailyBudget);
        vo.setMonthlyBudget(monthlyBudget);
        return vo;
    }

    /**
     * 创建预警状态
     */
    public static BudgetStatusVO warning(String message, BigDecimal todayCost, BigDecimal monthCost,
                                         BigDecimal dailyBudget, BigDecimal monthlyBudget) {
        BudgetStatusVO vo = new BudgetStatusVO();
        vo.setCanProceed(true);
        vo.setStatus("warning");
        vo.setMessage(message);
        vo.setTodayCost(todayCost);
        vo.setMonthCost(monthCost);
        vo.setDailyBudget(dailyBudget);
        vo.setMonthlyBudget(monthlyBudget);
        return vo;
    }

    /**
     * 创建超出状态
     */
    public static BudgetStatusVO exceeded(String message, BigDecimal todayCost, BigDecimal monthCost,
                                          BigDecimal dailyBudget, BigDecimal monthlyBudget) {
        BudgetStatusVO vo = new BudgetStatusVO();
        vo.setCanProceed(false);
        vo.setStatus("exceeded");
        vo.setMessage(message);
        vo.setTodayCost(todayCost);
        vo.setMonthCost(monthCost);
        vo.setDailyBudget(dailyBudget);
        vo.setMonthlyBudget(monthlyBudget);
        return vo;
    }
}

File: go-backend/internal/model/vo/budget_status_vo.go
// Package vo 预算状态视图对象
// @author <a href="https://codefather.cn">编程导航学习圈</a>
package vo

type BudgetStatusVO struct {
	CanProceed           bool    `json:"canProceed"`
	Status               string  `json:"status"`
	Message              string  `json:"message"`
	TodayCost            float64 `json:"todayCost"`
	MonthCost            float64 `json:"monthCost"`
	DailyBudget          float64 `json:"dailyBudget"`
	MonthlyBudget        float64 `json:"monthlyBudget"`
	DailyUsagePercent    float64 `json:"dailyUsagePercent"`
	MonthlyUsagePercent  float64 `json:"monthlyUsagePercent"`
}

const (
	BudgetStatusNormal   = "normal"
	BudgetStatusWarning  = "warning"
	BudgetStatusExceeded = "exceeded"
)

func BudgetStatusNormalVO(todayCost, monthCost, dailyBudget, monthlyBudget float64) BudgetStatusVO {
	return BudgetStatusVO{
		CanProceed:    true,
		Status:        BudgetStatusNormal,
		Message:       "预算充足",
		TodayCost:     todayCost,
		MonthCost:     monthCost,
		DailyBudget:   dailyBudget,
		MonthlyBudget: monthlyBudget,
	}
}

func BudgetStatusWarningVO(message string, todayCost, monthCost, dailyBudget, monthlyBudget float64) BudgetStatusVO {
	return BudgetStatusVO{
		CanProceed:    true,
		Status:        BudgetStatusWarning,
		Message:       message,
		TodayCost:     todayCost,
		MonthCost:     monthCost,
		DailyBudget:   dailyBudget,
		MonthlyBudget: monthlyBudget,
	}
}

func BudgetStatusExceededVO(message string, todayCost, monthCost, dailyBudget, monthlyBudget float64) BudgetStatusVO {
	return BudgetStatusVO{
		CanProceed:    false,
		Status:        BudgetStatusExceeded,
		Message:       message,
		TodayCost:     todayCost,
		MonthCost:     monthCost,
		DailyBudget:   dailyBudget,
		MonthlyBudget: monthlyBudget,
	}
}
